examples/misc/linalg.py

```diff
@@ -0,0 +1,134 @@
+    """
+    pass
+
+
+def option_basic(f):
+    def callback_shape(ctx, param, value):
+        return as_tuple(value)
+
+    def callback_opts(ctx, param, value):
+        if value is True:
+            return ('blocking,simd,openmp', {'blockalways': True, 'blockinner': True})
+        else:
+            return 'noop'
+
+    options = [
+        click.option('-ms', '--mat-shape', default=(4, 4), help='Matrix shape'),
+        click.option('-vs', '--vec-shape', default=4, help='Vector shape',
+                     callback=callback_shape),
+        click.option('-o', '--optimize', default=False, is_flag=True,
+                     help='Generate optimized code', callback=callback_opts)
+    ]
+    for option in reversed(options):
+        f = option(f)
+    return f
+
+
+@linalg.command(name='mat-vec')
+@option_basic
+def cli_mat_vec(mat_shape, vec_shape, optimize, **kwargs):
+    """``Ax = b``."""
+    i, j = dimensions('i j')
+    A = Function(name='A', shape=mat_shape, dimensions=(i, j))
+    x = Function(name='x', shape=vec_shape, dimensions=(j,))
+    b = Function(name='b', shape=vec_shape, dimensions=(i,))
+    mat_vec(A, x, b, optimize)
+
+
+@linalg.command(name='transpose')
+@option_basic
+def cli_transpose(mat_shape, optimize, **kwargs):
+    """``A -> A^T``."""
+    i, j = dimensions('i j')
+    A = Function(name='A', shape=mat_shape, dimensions=(i, j))
+    transpose(A, optimize)
+
+
+@linalg.command(name='transpose-mat-vec')
+@option_basic
+def cli_transpose_mat_vec(mat_shape, vec_shape, optimize, **kwargs):
+    """``A -> A^T, A^Tx = b``."""
+    i, j = dimensions('i j')
+    A = Function(name='A', shape=mat_shape, dimensions=(i, j))
+    x = Function(name='x', shape=vec_shape, dimensions=(j,))
+    b = Function(name='b', shape=vec_shape, dimensions=(i,))
+    transpose_mat_vec(A, x, b, optimize)
+
+
+@linalg.command(name='mat-mat')
+@option_basic
+def cli_mat_mat(mat_shape, optimize, **kwargs):
+    """``AB = C``."""
+    i, j, k = dimensions('i j k')
+    A = Function(name='A', shape=mat_shape, dimensions=(i, j))
+    B = Function(name='B', shape=mat_shape, dimensions=(j, k))
+    C = Function(name='C', shape=mat_shape, dimensions=(i, k))
+    mat_mat(A, B, C, optimize)
+
+
+@linalg.command(name='mat-mat-sum')
+@option_basic
+def cli_mat_mat_sum(mat_shape, optimize, **kwargs):
+    """``AB + AC = D``."""
+    i, j, k = dimensions('i j k')
+    A = Function(name='A', shape=mat_shape, dimensions=(i, j))
+    B = Function(name='B', shape=mat_shape, dimensions=(j, k))
+    C = Function(name='C', shape=mat_shape, dimensions=(j, k))
+    D = Function(name='D', shape=mat_shape, dimensions=(i, k))
+    mat_mat_sum(A, B, C, D, optimize)
+
+
+def mat_vec(A, x, b, optimize):
+    """``Ax = b``."""
+    op = Operator(Eq(b, A*x), dle=optimize)
+    op.apply()
+    info('Executed `Ax = b`')
+
+
+def transpose(A, optimize):
+    """``A -> A^T``."""
+    i, j = A.indices
+    op = Operator(Eq(A, A.indexed[j, i]), dle=optimize)
```

Comment:
Don't think this works. In place transpose of a matrix are really tricky.
A[i, j] = A[j,.i]

so when reaches A[j, i] can't replace by A[i, j] anymore.

Can be tested with `np.dot(mat_vec(A,x), y) == np.dot(mat_vec(A^T,y),x)`

Reply:
dumb me